mqtt/mqtt-bridge/src/persist/waking_state/waking_map.rs

```diff
@@ -0,0 +1,57 @@
+#![allow(dead_code)] // TODO remove when ready
+
+use std::{cmp::min, collections::HashMap, collections::VecDeque, task::Waker};
+
+use mqtt3::proto::Publication;
+
+use crate::persist::{waking_state::StreamWakeableState, Key, PersistError};
+
+/// When elements are retrieved they are moved to the in flight collection.
+pub struct WakingMap {
+    queue: VecDeque<(Key, Publication)>,
+    in_flight: HashMap<Key, Publication>,
+    waker: Option<Waker>,
+}
+
+impl WakingMap {
+    pub fn new() -> Self {
+        let queue: VecDeque<(Key, Publication)> = VecDeque::new();
+        let in_flight = HashMap::new();
+
+        WakingMap {
+            queue,
+            in_flight,
+            waker: None,
+        }
+    }
+}
+
+impl StreamWakeableState for WakingMap {
+    fn insert(&mut self, key: Key, value: Publication) -> Result<(), PersistError> {
+        self.queue.push_back((key, value));
+
+        if let Some(waker) = self.waker.take() {
+            waker.wake();
+        }
+
+        Ok(())
+    }
+
+    fn batch(&mut self, count: usize) -> Result<Vec<(Key, Publication)>, PersistError> {
+        let count = min(count, self.queue.len());
+        let output: Vec<_> = self.queue.drain(..count).collect();
+        self.in_flight.extend(output.clone().into_iter());
```

Comment:
this looks suspicious. why all items from the next batch is automatically going to inflight?

Reply:
usually inflilght queue is 16, but batch size can be 100

Reply:
Yes good point. They should only go in flight once actually returned by the loader. So by this:
> usually inflilght queue is 16, but batch size can be 100

Do you mean that we should fail a loader.next() if we have 16 already in flight?

Reply:
I think changing this will have negative perf implications. Previously, the could avoid many locks of the underlying state because it could process off of its current batch (which was already marked in-flight). If we change to add to in-flight only once loaded, then the loader has to lock the shared state every time to see if in_flight has enough remaining capacity. So we can either:

**Option A**
Keep the current approach. This will have better perf, but has less stringent enforcement of what can be removed from the underlying state. For example, it is possible to remove an element once it is in the loader's next batch, but not returned by a ```loader.next()``` call. 

With this approach, batch size and in-flight size are the same.

**Option B**
Switch to remove in-flight only once returned from the loader. This will require locking the shared state to check the in-flight size on every ```loader.next()``` call.

With this approach, batch size and in-flight size can be configured differently. However, batch size becomes less meaningful because it won't save locks. Previously, we only had to lock when we got a new batch. Now we have to lock every time to check the in-flight size.